boot/seal.go

```diff
@@ -25,65 +25,216 @@ import (
+	}
+
+	recoveryBootChain, err := buildRecoveryBootChain(rbl, model, modeenv)
+	if err != nil {
+		return fmt.Errorf("cannot build recovery boot chain: %v", err)
+	}
+
+	// build the run mode boot chains
+	bl, err := bootloader.Find(InitramfsUbuntuBootDir, &bootloader.Options{
+		Role:        bootloader.RoleRunMode,
+		NoSlashBoot: true,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot find the bootloader: %v", err)
+	}
 	cmdline, err := ComposeCandidateCommandLine(model)
 	if err != nil {
-		return fmt.Errorf("cannot obtain kernel command line: %v", err)
+		return fmt.Errorf("cannot compose the candidate command line: %v", err)
 	}
 
-	// Get the expected kernel command line of the recovery system we're installing from
-	recoveryCmdline, err := ComposeRecoveryCommandLine(model, modeenv.RecoverySystem)
+	runModeBootChains, err := buildRunModeBootChains(rbl, bl, model, modeenv, cmdline)
 	if err != nil {
-		return fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+		return fmt.Errorf("cannot build run mode boot chain: %v", err)
 	}
 
-	kernelCmdlines := []string{
-		cmdline,
-		recoveryCmdline,
+	pbc := toPredictableBootChains(append(runModeBootChains, *recoveryBootChain))
+
+	roleToBlName := map[bootloader.Role]string{
+		bootloader.RoleRecovery: rbl.Name(),
+		bootloader.RoleRunMode:  bl.Name(),
 	}
 
-	sealKeyParams := secboot.SealKeyParams{
-		ModelParams: []*secboot.SealKeyModelParams{
-			{
-				Model:          model,
-				KernelCmdlines: kernelCmdlines,
-				EFILoadChains:  []*secboot.LoadChain{loadChain},
-			},
-		},
-		KeyFile:                 filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
-		TPMPolicyUpdateDataFile: filepath.Join(InstallHostFDEDataDir, "policy-update-data"),
-		TPMLockoutAuthFile:      filepath.Join(InstallHostFDEDataDir, "tpm-lockout-auth"),
+	// get parameters from bootchains and seal the key
+	params, err := sealKeyParams(pbc, roleToBlName)
+	if err != nil {
+		return fmt.Errorf("cannot build key sealing parameters: %v", err)
 	}
 
-	if err := secbootSealKey(key, &sealKeyParams); err != nil {
+	if err := secbootSealKey(key, params); err != nil {
 		return fmt.Errorf("cannot seal the encryption key: %v", err)
 	}
 
+	// TODO:UC20: store the predictable bootchains
+
 	return nil
 }
+
+func buildRecoveryBootChain(rbl bootloader.Bootloader, model *asserts.Model, modeenv *Modeenv) (bc *bootChain, err error) {
+	tbl, ok := rbl.(bootloader.TrustedAssetsBootloader)
+	if !ok {
+		return nil, fmt.Errorf("bootloader doesn't support trusted assets")
+	}
+
+	// get the command line
+	cmdline, err := ComposeRecoveryCommandLine(model, modeenv.RecoverySystem)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+	}
+
+	// get kernel information from seed
+	perf := timings.New(nil)
+	_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, modeenv.RecoverySystem, []snap.Type{snap.TypeKernel}, perf)
```

Comment:
using modeenv.RecoverySystem make sense only for install-time seal, in general we need to use modeenv.CurrentRecoverySystems (which is a list in principle)